[Wang Zihao] iP

9teMare opened the pull request:
## **CortanaTodo**
> “Productivity is being able to do things that you were never able to do before.” – Franz Kafka ([source](https://dansilvestre.com/productivity-quotes/))

### CortanaTodo frees your mind by keeping track of the things that you need to do, it's
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

### All you need to do is,
 1. Download it from [here](https://github.com/9teMare/ip/releases/tag/A-Jar).
 2. Double click it.
 3. Add your tasks.
 4. Free your mind, let Cortana manage your tasks for you 😆
 
 And it's FREE!

### Features:
- [x] Managing todos
- [x] Managing deadlines
- [x] Managing events
- [x] Find tasks on the same date/time
- [x] Find tasks with keyword

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
    String PATH = Paths.get("").toAbsolutePath() + "/data/";
    new Duke(PATH).run();
}
```

Commit:
Add Gradle support

9teMare committed:
Completed Level-1, renamed chatbot to Cortana.

9teMare committed:
Completed Level-2, added some additional features.

9teMare committed:
Renamed method, refactored code

9teMare committed:
Completed Level-3

Used OOP, added Task class and TaskManager class.

9teMare committed:
Level-3 completed without using recommended extensions.

9teMare committed:
Revert "Level-4 completed without using recommended extensions."

This reverts commit 40be96b8c573c818d546870893686554af3a352e.

9teMare committed:
Level-4 completed without using recommended extensions.

9teMare committed:
Completed Level-4 with extensions

Refactored code, removed TaskManager, modified Duke, Task and added Event, Deadline, Todo.

9teMare committed:
Completed Level-5

Completed Level-5, refactored code, but still need additional edits.

9teMare committed:
Completed Level-6, rewrote Level-5

9teMare committed:
Used enums, correctly implemented taskSet to prevent duplicate of tasks

9teMare committed:
Separated greetings

9teMare committed:
Cleaned up code, removed enums, passed all the tests.

9teMare committed:
Attempted to add level-7 on top of master in branch-level-7

9teMare committed:
Completed Level-7

9teMare committed:
Completed Level-8

9teMare committed:
Merge branch 'master' into branch-Level-7

9teMare committed:
Merge pull request #1 from 9teMare/branch-Level-7

Branch level 7

9teMare committed:
Merge pull request #2 from 9teMare/branch-Level-8

Completed Level-8

9teMare committed:
Add the feature to show all tasks in the same date/time.

9teMare committed:
Completed A-MoreOOP and A-Packages, haven't thoroughly tested

Largely refactored code.

9teMare committed:
Made fields in Duke.java final

9teMare committed:
Change public to private in Deadline and Event, added setters. Starting to implement JUnit test.

9teMare committed:
Refactor code, completed minimum requirement for A-JUnit

9teMare committed:
Add Javadoc

9teMare committed:
gitignore Manifest.mf

9teMare committed:
Complete level-9

9teMare committed:
Refactor to comply with coding standard

9teMare committed:
Merge pull request #3 from 9teMare/branch-A-JavaDoc

Add Javadoc

9teMare committed:
Gitignore task.txt

9teMare committed:
Merge branch 'branch-A-CodingStandard'

9teMare committed:
Add the missing space for CortanaException message in Parser line 98

9teMare committed:
Gitignore task.txt

9teMare committed:
Resolve conflict in gitignore

9teMare committed:
Merge branch 'branch-Level-9'

9teMare committed:
Add JUnit test for FindCommand

9teMare committed:
Merge pull request #6 from 9teMare/add-gradle-support

Add Gradle support

9teMare committed:
Refactor code to pass gradle checkStyle

9teMare committed:
Setup Gradle to build, run and test, fix coding style violations with checkStyle

9teMare committed:
Implement GUI

tzhan98 commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,38 @@
+    private final Duke duke = new Duke(Paths.get("").toAbsolutePath() + "/data/");
+
+    @Override
+    public void start(Stage stage) {
```
Missing javaDoc

tzhan98 commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,12 @@
+import javafx.application.Application;
+
+/**
+ * A launcher class to workaround classpath issues.
```
Just curious but what classpath issue was encountered and how does this class overcome it?
Maybe you could summarize that in your JavaDoc too for readers to better understand?

tzhan98 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,69 @@
+    /**
+     * The program is not yet exited.
+     */
+    public boolean isExit() {
```
~~Should it be named `isExit` if it always returns false?
Would a more suitable function name be `exitNow`?~~
Read below and found that current `isExit()` overrides the abstract method and there are multiple implementation of these

tzhan98 replied:
Maybe you could annotate `@Override` so that it flags if you are not overriding the method in `Command` and readers would know there could be other implementation of `isExit()` 😃

tzhan98 commented on src/main/java/duke/command/ShowAllTasksOnSameDateCommand.java:
```diff
@@ -0,0 +1,118 @@
+     * @param ui the ui to operate on
+     * @param storage the storage to operate on
+     */
+    public String execute(TaskList taskList, Ui ui, Storage storage) throws CortanaException {
```
Would this be a '[deep nesting](https://nus-cs2103-ay2122s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-deep-nesting)' ?

cpwill01 replied:
I agree with @tzhan98. Perhaps you could consider the suggestions given [here](https://blog.codinghorror.com/flattening-arrow-code/)?

tzhan98 commented on src/main/java/duke/command/ShowAllTasksOnSameDateCommand.java:
```diff
@@ -0,0 +1,118 @@
+import duke.utils.CortanaException;
+import duke.utils.Storage;
+import duke.utils.TaskList;
+import duke.utils.Ui;
```
Great that you sorted your import statements!

tzhan98 commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,195 @@
+     * @return the specific Cortana command
+     * @throws CortanaException the cortana exception
+     */
+    public static Command parse(String input) throws CortanaException {
```
Maybe you could shorten this method? Seems like a [long method](https://nus-cs2103-ay2122s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods)

tzhan98 reviewed (commented):
LGTM.
- Pretty cool that you customized it to look like Cortana (the loading picture too)
- Just a few minor suggestions
- Liked it that you broke down ~~duke~~ Cortana to many different classes
- The code followed the style and code quality guide strictly

Keep up the good work! 👍 💯

cpwill01 commented on src/test/java/utils/ParserTest.java:
```diff
@@ -0,0 +1,159 @@
+public class ParserTest {
+    @Test
+    @DisplayName("Should parse to ExitCommand")
+    public void parsedToExitCommand() {
```
Should these tests be renamed to the format of `featureUnderTest_testScenario_expectedBehavior()` as specified [here](https://se-education.org/guides/conventions/java/intermediate.html#naming)?

cpwill01 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,69 @@
+     * @param taskList the task list to operate on
+     * @param ui the ui to operate on
+     * @param storage the storage to operate on
+     */
```
Should there be an `@throws` here in the Javadoc (as well as in the other command classes)?

cpwill01 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    /**
+     * Execute the add task operation.
```
I think this should start with "Executes" rather than "Execute".

cpwill01 replied:
I also noticed similar issues with the Javadoc header in the other Command subclasses, as well as the Storage class and UI class.

cpwill01 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    /**
+     * The program is not yet exited.
```
Would it be possible to rephrase the header comment here (as well as in the other command classes) so that it [starts with a singular form verb](https://se-education.org/guides/conventions/java/intermediate.html#comments)? eg. "Indicates" or "Returns"

cpwill01 commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,50 @@
+     */
+    public String execute(TaskList taskList, Ui ui, Storage storage) {
+        StringBuilder tasksToList = new StringBuilder();
+        //String tasks = "";
```
Should this line and line 28 be removed?

cpwill01 reviewed (commented):
I really like the customisations you made to the chatbot! 😄 I found your code easy to read overall, except at places where there was some deep nesting. Other than that, just a few nits to fix in the Javadoc header comments for the Command subclasses and Storage and UI classes.

9teMare committed:
Fix load from file parse time error

9teMare committed:
Fix some deep nesting issue

9teMare committed:
Implement assertions

Currently there is no assertions to ensure certain assumptions hold.

Need to have assertion to make sure that the author, which is myself, will not break certain part of the code when those assumptions are reached.

I therefore implement assertions to the following parts:
- In AddCommand, assert to ensure that all the task being added are instance of either Deadline, Event or Todo
- In DeleteAllCommand, assert to ensure that after deletion, there is no tasks left in the taskList and taskSet, also to ensure the file is empty.
- In DeleteCommand, assert to ensure that after deletion, the task needed to be removed is no longer present in the taskList and taskSet.
- In FindCommand, assert to ensure that the search keyword is not empty.
- In MarkCommand, UnmarkCommand and DeleteCommand, assert to ensure that after exceptions are caught correctly
- In Duke, assert to ensure that loadFile is not null.
- In Storage, assert to ensure that the tasks from the file are instance of either Deadline, Event or Todo. Also to ensure that when there is a need to write to the file, there is actually something to be written instead of writing something empty.
- In Task, assert to ensure that the description is not empty.

Other than assertions, commented out codes are removed.

9teMare committed:
Refactor Storage method

Currently there are some duplicated code inside Storage.

Need to refactor to ensure better code quality. This is the beginning of the refactor process.

Also added an enum called TaskType, will see if it will be useful in future refactors.

9teMare committed:
Refactor code to enhance code quality

Currently, some of the codes have deep nesting issue, and some of the methods are too long.

Need to resolve the deep nesting issue effectively in order to ensure better readability, and also to make the code easier to debug. Methods that are too long and have duplicate codes are shortened, some common functionalities are extracted out to stand alone methods.

Other than code refactoring, the following are done in this commit:
- Removed redundant methods that are not required anymore, i.e. isExit() method in command package
- Added enum along with switch case to resolve deep nesting issue.
- Added missing JavaDoc

9teMare committed:
Implement stream on certain methods

Currently, the execute() methods in FindCommand and ShowAllTasksOnSameDateCommand are implemented using for loops.

The current implementation introduces lengthy codes, with the implementation of stream, codes are drastically shortened, which increases readability.

9teMare committed:
Merge pull request #7 from 9teMare/branch-A-Assertions

Implement assertions

9teMare committed:
Merge branch 'master' into branch-A-CodeQuality

9teMare committed:
Merge pull request #8 from 9teMare/branch-A-CodeQuality

Branch a code quality

9teMare committed:
Merge branch 'master' into branch-A-Streams

9teMare committed:
Merge pull request #9 from 9teMare/branch-A-Streams

Implement stream on certain methods

9teMare committed:
Fix bug in Ui where extra characters are printed

9teMare committed:
Create gradle.yml to enable CI

9teMare committed:
Add missing JavaDoc for FindCommand

9teMare committed:
Add view schedules feature

Choose B-ViewSchedules as the extension to implement.
C-DetectDuplicates was already implemented a few weeks ago.

9teMare committed:
Add Missing JavaDoc for ViewSchedulesCommand

9teMare committed:
Add support for mark/unmark/delete more than 1 task at the same time through varargs

9teMare committed:
Remove extra blank space in Ui.java

9teMare committed:
Fix corresponding bug in UiTest.java

9teMare committed:
Improve coding standard in Parser

Drastically shortened the parse() method by applying the SLAP principle.

9teMare committed:
Implement error message recoloring

- Now when Cortana responses with error messages, the responses will be in red color
- Add an isReturnError attribute in Cortana
- Fix some regex bugs

9teMare committed:
Update user guide

9teMare committed:
Set theme jekyll-theme-minimal

9teMare committed:
Set theme jekyll-theme-architect

9teMare committed:
Update README.md

9teMare committed:
Set theme jekyll-theme-architect

9teMare committed:
Update README.md

9teMare committed:
Update the jar name in build.gradle
